api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java

```diff
@@ -691,9 +713,91 @@ public String getSecretQuestion(User user) throws APIException {
 		if (user.getUserId() != null) {
 			LoginCredential loginCredential = dao.getLoginCredential(user);
 			return loginCredential.getSecretQuestion();
-		} else {
+		} 
+		else {
 			return null;
 		}
     }
 	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByUsernameOrEmail(java.lang.String)
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByUsernameOrEmail(String usernameOrEmail) {
+		if (StringUtils.isNotBlank(usernameOrEmail)) {
+			User user = dao.getUserByEmail(usernameOrEmail);
+			if (user == null) {
+				return getUserByUsername(usernameOrEmail);
+			}
+			return user;
+		}
+		throw new APIException("error.usernameOrEmail.notNullOrBlank", (Object[]) null);
+	}
+	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByActivationKey(java.lang.String)
+	 * 
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByActivationKey(String activationKey) {
+		LoginCredential loginCred = dao.getLoginCredentialByActivationKey(activationKey);
+		if (loginCred != null) {
+			String[] credTokens = loginCred.getActivationKey().split(":");
+			if (System.currentTimeMillis() <= Long.parseLong(credTokens[1])) {
```

Comment:
What happens if the credTokens length is 0?

Reply:
credTokens length cannot be null at this level because if loginCred is not null then there is an activation key of the form xddkkdud7i8d:55636737352 which ensures that credToken lenght is always greater than 0, other wise  null will be returned instead of a failure